### PR TITLE
[CPROD-1196] Stable storage

### PR DESCRIPTION
**Pull request by F3kilo**
Changes:
- `Metadata` stored in `Cell` from `ic-stable-structures` and removed from the `CanisterState`. Now it works with storage semantics providing `get()/set()` methods. I think automatic stable memory update, using variable semantic `metadata = other_metadata`, is not necessary.
- `Balances` stored in `StableMultimap` from `ic-stable-structures` and removed from the `CanisterState`.  Also, `Principal`, `Subaccount` and amount (`u128`), are stored in stable memory as raw bytes now. The `StableMultimap` require to set max bytes length for key and for value. Its a little tricky to determine that for candid and protobuf protocols. If we want to use some (de)serialization protocol, we can make it in other ticket. API of storage don't depend on it.
- `CanisterState` removed and all code like `self.state.borrow().balances` replaces with direct usage of `StableBalances`.
- Canister `Stats` renamed to `TokenConfig`.

### CODE REVIEW

File: src/token/api/src/state/balances.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        let principal_len = bytes[0] as usize;
+        let account = Principal::from_slice(&bytes[1..principal_len + 1]);
+        let mut subaccount = [0u8; SUBACCOUNT_LEN];
+        subaccount.copy_from_slice(&bytes[SUBACCOUNT_OFFSET..]);
+        Self {
+            principal: account,
+            subaccount,
+        }
+    }
+}
```
Comment (Maximkaaa) [inline src/token/api/src/state/balances.rs:224]: Let's add a couple unit tests for this. This is too scary to leave without dedicated unit tests.

File: src/token/api/src/storage/structures.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+    /// Updates value in stable memory.
+    pub fn set(&mut self, value: T) {
+        let canister_id = ic::id();
+        match self.data.entry(canister_id) {
+            Entry::Occupied(mut entry) => {
+                entry
+                    .get_mut()
+                    .set(value)
+                    .expect("failed to set value to stable cell");
```
Comment (Maximkaaa) [inline src/token/api/src/storage/structures.rs:40]: Return an error, do not panic here. You don't know at what point this code is executed and if it's ok to panic there or not.

File: src/token/api/src/storage/structures.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            Entry::Vacant(entry) => {
+                let memory = super::get_memory_by_id(self.memory_id);
+                entry.insert(cell::Cell::init(memory, value).expect("failed to init stable cell"));
```
Comment (Maximkaaa) [inline src/token/api/src/storage/structures.rs:44]: Do not panic here also

File: src/token/api/src/storage/structures.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    pub fn insert(&mut self, key: K, value: V) {
+        let canister_id = ic::id();
+        self.data
+            .entry(canister_id)
+            .or_insert_with(|| {
+                let memory = super::get_memory_by_id(self.memory_id);
+                btreemap::BTreeMap::init(memory, self.max_key_size, self.max_value_size)
+            })
+            .insert(key, value)
+            .expect("failed to insert value to stable btree map");
```
Comment (Maximkaaa) [inline src/token/api/src/storage/structures.rs:82]: Do not panic

File: src/token/api/src/state/ledger.rs [hunk 5]
```diff
@@ -1,14 +1,102 @@
...
+        Self::with_ledger(|ledger| ledger.claim(claim_account, to, amount))
+    }
+
+    fn with_ledger<F, R>(f: F) -> R
+    where
+        F: FnOnce(&mut Ledger) -> R,
+    {
+        LEDGER.with(|ledgers| {
+            let canister_id = ic::id();
+            println!("canis id: {canister_id}");
```
Comment (Maximkaaa) [inline src/token/api/src/state/ledger.rs:92]: del

**Review by Maximkaaa — CHANGES_REQUESTED** [review]
Overall looks fine, should work as expected. A few architectural considerations I sent as a DM. One problem with the implementation though is since we don't store the ledger in the stable memory, after upgrade TxIds will start again from 0. This has security risks, as other applications may rely on the ids to be unique. So at least we must store the last tx id somewhere in the stable memory.

**Review by hagsteel — CHANGES_REQUESTED** [review]
looks like I've already done a review on this but none of the comments has been addressed?

File: src/token/api/src/canister/icrc1_transfer.rs [hunk 6]
```diff
@@ -687,13 +721,17 @@ mod tests {
...
+
         let history_size_before = canister.history_size();
 
         ctx.update_caller(john());
-        canister
-            .burn(None, None, Tokens128::from(1001))
-            .unwrap_err();
+        println!("john: {}", john());
+        println!("alice: {}", alice());
+        dbg!(canister.burn(None, None, Tokens128::from(1001))).unwrap_err();
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply src/token/api/src/canister/icrc1_transfer.rs]: Done

File: src/token/api/src/canister/icrc1_transfer.rs [hunk 7]
```diff
@@ -1124,7 +1162,11 @@ mod proptests {
             // pass, because since we are running auction state on each
             // endpoint call, it affects `BiddingInfo.fee_ratio` that is
             // used for charging fees in `approve` endpoint.
-            canister.state.borrow_mut().stats.min_cycles = 0;
+
+        let mut stats = TokenConfig::get_stable();
+            stats.min_cycles = 0;
+        TokenConfig::set_stable(stats);
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply src/token/api/src/canister/icrc1_transfer.rs]: Done

File: src/token/api/src/state/balances.rs [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            .filter(|(account, _)| account.owner == owner)
+            .map(|(account, amount)| (account.subaccount, amount))
+            .collect()
+    }
+
+    /// Return sum of all balances.
+    fn total_supply(&self) -> Tokens128 {
+        self.list_balances(0, usize::MAX)
+            .into_iter()
+            .fold(Tokens128::ZERO, |a, b| (a + b.1).unwrap_or(Tokens128::ZERO))
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply src/token/api/src/state/balances.rs]: Done, but Jonas will not like it)

File: src/token/api/src/state/balances.rs [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    }
+
+    fn remove(&mut self, account: &AccountInternal) -> Option<Tokens128> {
+        self.0.remove(account)
+    }
+
+    fn total_supply(&self) -> Tokens128 {
+        self.0
+            .iter()
+            .fold(Tokens128::ZERO, |a, b| (a + b.1).unwrap_or(Tokens128::ZERO))
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply src/token/api/src/state/balances.rs]: Done

File: src/token/api/src/state/balances.rs (hunk 1 above)
Comment (F3kilo) [reply src/token/api/src/state/balances.rs:230]: Done

File: src/token/api/src/storage/structures.rs (hunk 2 above)
Comment (F3kilo) [reply src/token/api/src/storage/structures.rs]: Done. Now expects are in client code, which knows that the only reason for errors is out of stable memory, which we can't handle somehow.

File: src/token/api/src/storage/structures.rs (hunk 3 above)
Comment (F3kilo) [reply src/token/api/src/storage/structures.rs]: Done

File: src/token/api/src/storage/structures.rs (hunk 4 above)
Comment (F3kilo) [reply src/token/api/src/storage/structures.rs]: Done

File: src/token/api/src/state/ledger.rs (hunk 5 above)
Comment (F3kilo) [reply src/token/api/src/state/ledger.rs]: Done

File: src/token/api/src/canister/is20_transactions.rs [hunk 10]
```diff
@@ -319,43 +318,52 @@ pub fn batch_transfer(
...
-    updated_balances.set_balance(auction_account(), balances.balance_of(auction_account()));
+    let auction_acc = auction_account();
+
+    let mut updates = LocalBalances::from_iter(
+        [
+            (from, balances.balance_of(&from)),
+            (fee_to, balances.balance_of(&fee_to)),
+            (auction_acc, balances.balance_of(&auction_acc)),
+        ]
+        .into_iter(),
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply src/token/api/src/canister/is20_transactions.rs]: Done

File: src/token/api/src/canister/is20_transactions.rs [hunk 11]
```diff
@@ -365,13 +373,13 @@ pub(crate) fn batch_transfer_internal(
...
             TxError::InsufficientFunds { .. } => TxError::InsufficientFunds {
-                balance: balances.balance_of(from),
+                balance: balances.balance_of(&from),
             },
             other => other,
         })?;
     }
 
-    balances.apply_change(&updated_balances);
+    balances.apply_updates(updates.list_balances(0, usize::MAX).into_iter());
```
[the remark this answers is not in the document]
Comment (F3kilo) [reply src/token/api/src/canister/is20_transactions.rs]: Sure, its better. Done

**Review by hagsteel — APPROVED** [review]

**Comment by F3kilo** [comment]
We decided to move memory manager and stable structures wrappers to the `canister-sdk` repo, because we already have three canisters, which will use that code:
- Is20,
- Pair,
- EVM.

File: .gitignore [hunk 12]
```diff
@@ -8,4 +8,6 @@ token-factory.did
 token.wasm
 /proptest-regressions
 
-/.DS_Store
\ No newline at end of file
+/.DS_Store
+
+Cargo.lock
```
Comment (Maximkaaa) [inline .gitignore:13]: You added this to .gitignore, but the file is still commited. It should be either removed from git, or from .gitignore.

**Review by Maximkaaa — CHANGES_REQUESTED** [review]
I believe we still have an issue for TX ids after upgrades. Current implementation saves the tx offset to stable memory, but only when the history is trimmed. So all transactions that are still in-memory will be lost togther with their IDs, so the ids will still be duplicated after upgrade.

To fix this issue fully we need to write last tx id to the stable memory on every push to the ledger.

File: .gitignore (hunk 12 above)
Comment (F3kilo) [reply .gitignore]: Removed it from `.gitignore`.

**Comment by F3kilo** [comment]
> To fix this issue fully we need to write last tx id to the stable memory on every push to the ledger.

Now I store total tx count in stable memory, updating it for each new tx. Index calculation logic changed accordingly.

**Review by Maximkaaa — APPROVED** [review]

**Pull request merged**